### PR TITLE
feat(ingestion): conversion buffer producer

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

#9182 

## Changes

This should be a no-op change for now given the buffer is disabled.

It splits out the event creation from the `processEvent` path in the worker so that we can selectively create an event or not. If the event shouldn't be immediately created, it will be produced to the buffer, which will in turn trigger ingestion at a later point.

## How did you test this code?

Added tests

### CODE REVIEW

**Review by tiina303 — APPROVED** [review]

File: plugin-server/src/types.ts [hunk 1]
```diff
@@ -879,3 +881,13 @@ export enum PluginServerMode {
...
 }
+
+export interface PreIngestionEvent {
+    eventUuid: string
+    event: string
+    teamId: TeamId
+    distinctId: string
+    properties: Properties
+    timestamp: DateTime | string
+    elementsList: Element[]
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/types.ts:892]: this is used in a lot of places.

before we used to pass all these as args to createEvent. Now however we need to return these, send these to the buffer, etc. I could check other types though and see if there's significant overlap to reuse

File: plugin-server/src/worker/ingestion/ingest-event.ts [hunk 2]
```diff
@@ -60,3 +62,47 @@ export async function ingestEvent(hub: Hub, event: PluginEvent): Promise<IngestE
...
+    elements?: Element[],
+    person?: Person
+): Promise<Action[]> {
+    let actionMatches: Action[] = []
+
+    actionMatches = await hub.actionMatcher.match(event, person, elements)
+    await hub.hookCannon.findAndFireHooks(event, person, siteUrl, actionMatches)
+
+    // eventId is undefined for CH deployments
+    // CH deployments calculate actions on the fly
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/ingest-event.ts:87]: this comment already existed but sure can remove

File: plugin-server/src/worker/ingestion/ingest-event.ts [hunk 3]
```diff
@@ -60,3 +62,47 @@ export async function ingestEvent(hub: Hub, event: PluginEvent): Promise<IngestE
...
+    if (actionMatches.length && eventId !== undefined) {
+        await hub.db.registerActionMatch(eventId, actionMatches)
+    }
+
+    return actionMatches
+}
+
+// TODO: Handle new persons?
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/ingest-event.ts:95]: So the rules for the buffer are specified here:

#9182 

One of the rules is that events for persons that didn't exist before should go to the buffer. However, while building this, I realized:

1. It'd be a major refactor to cover that
2. The conversion seconds anyway handles it (person created within 60sec)

The one edge case is if we lower that buffer value a lot (say to 1sec), but actually I've decided to punt on this problem

File: plugin-server/src/worker/ingestion/ingest-event.ts [hunk 4]
```diff
@@ -60,3 +62,47 @@ export async function ingestEvent(hub: Hub, event: PluginEvent): Promise<IngestE
...
+function shouldSendEventToBuffer(hub: Hub, event: PreIngestionEvent, person?: Person) {
+    const isAnonymousEvent =
+        event.properties && event.properties['$device_id'] && event.distinctId === event.properties['$device_id']
+    const isRecentPerson = !person || DateTime.now().diff(person.created_at).seconds < hub.BUFFER_CONVERSION_SECONDS
+    const ingestEventDirectly = isAnonymousEvent || event.event === '$identify' || !isRecentPerson
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/ingest-event.ts:100]: We thought about the buffer a _lot_ before making decisions on the rules. Would have been worth questioning those earlier.

But anyway, no, that doesn't make sense. We will merge on the anonymous person ID, so the ID for those events will never change. The buffer is used for events where the person ID might change.

The reason for sending anonymous events there would be if we merged into the identified person ID, but under that the whole system crumbles. How long does it take for an anonymous user to become identified? Is there a guarantee that they will?

File: plugin-server/src/worker/ingestion/ingest-event.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiina303) [reply plugin-server/src/worker/ingestion/ingest-event.ts]: commented on that issue, sorry I missed it somehow before

File: plugin-server/src/worker/ingestion/ingest-event.ts [hunk 5]
```diff
@@ -60,3 +62,48 @@ export async function ingestEvent(hub: Hub, event: PluginEvent): Promise<IngestE
...
+}
+
+// context: https://github.com/PostHog/posthog/issues/9182
+// TL;DR: events from a recently created non-anonymous person are sent to a buffer
+// because their person_id might change. We merge based on the person_id of the anonymous user
+// so ingestion is delayed for those events to increase our chances of getting person_id correctly
+function shouldSendEventToBuffer(hub: Hub, event: PreIngestionEvent, person?: Person) {
+    const isAnonymousEvent =
+        event.properties && event.properties['$device_id'] && event.distinctId === event.properties['$device_id']
+    const isRecentPerson = !person || DateTime.now().diff(person.created_at).seconds < hub.BUFFER_CONVERSION_SECONDS
```
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/ingest-event.ts:100]: @tiina303

File: plugin-server/src/worker/ingestion/ingest-event.ts [hunk 6]
```diff
@@ -60,3 +62,48 @@ export async function ingestEvent(hub: Hub, event: PluginEvent): Promise<IngestE
         clearTimeout(timeout)
     }
 }
+
+export async function ingestBufferEvent(hub: Hub, event: PreIngestionEvent): Promise<IngestEventResponse> {
+    const person = await hub.db.fetchPerson(event.teamId, event.distinctId)
```
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/ingest-event.ts:67]: @tiina303 essentially I'd need to look at all the code paths and see what they use from person but we might be able to substitute this for a cached version

File: plugin-server/src/worker/ingestion/ingest-event.ts (hunk 5 above)
Comment (tiina303) [reply plugin-server/src/worker/ingestion/ingest-event.ts:100]: btw we also have https://github.com/PostHog/posthog/blob/c776ee3583a62b4ef0babdb7a5e192f3d45c07be/plugin-server/src/worker/ingestion/person-manager.ts#L18

**Pull request merged**